solidity/contracts/bridge/Bridge.sol

```diff
@@ -32,89 +36,233 @@ pragma solidity 0.8.4;
+    ///        P2(W)SH BTC deposit transaction,
     ///      - BTC deposit for the given `fundingTxHash`, `fundingOutputIndex`
-    ///        can be revealed by `msg.sender` only one time.
+    ///        can be revealed only one time.
     ///
     ///      If any of these requirements is not met, the wallet _must_ refuse
     ///      to sweep the deposit and the depositor has to wait until the
     ///      deposit script unlocks to receive their BTC back.
     function revealDeposit(
-        bytes32 fundingTxHash,
-        uint8 fundingOutputIndex,
-        uint64 blindingFactor,
-        bytes calldata refundPubKey,
-        uint64 amount,
-        address vault
+        TxInfo calldata fundingTx,
+        RevealInfo calldata reveal
     ) external {
-        uint256 depositId =
-            uint256(
-                keccak256(
-                    abi.encode(fundingTxHash, fundingOutputIndex, msg.sender)
-                )
+        bytes memory expectedScript =
+            abi.encodePacked(
+                hex"14", // Byte length of depositor Ethereum address.
+                reveal.depositor,
+                hex"75", // OP_DROP
+                hex"08", // Byte length of blinding factor value.
+                reveal.blindingFactor,
+                hex"75", // OP_DROP
+                hex"76", // OP_DUP
+                hex"a9", // OP_HASH160
+                hex"14", // Byte length of a compressed Bitcoin public key hash.
+                reveal.walletPubKeyHash,
+                hex"87", // OP_EQUAL
+                hex"63", // OP_IF
+                hex"ac", // OP_CHECKSIG
+                hex"67", // OP_ELSE
+                hex"76", // OP_DUP
+                hex"a9", // OP_HASH160
+                hex"14", // Byte length of a compressed Bitcoin public key hash.
+                reveal.refundPubKeyHash,
+                hex"88", // OP_EQUALVERIFY
+                hex"04", // Byte length of refund locktime value.
+                reveal.refundLocktime,
+                hex"b1", // OP_CHECKLOCKTIMEVERIFY
+                hex"75", // OP_DROP
+                hex"ac", // OP_CHECKSIG
+                hex"68" // OP_ENDIF
+            );
+
+        bytes memory fundingOutput =
+            fundingTx.outputVector.extractOutputAtIndex(
+                reveal.fundingOutputIndex
+            );
+        bytes memory fundingOutputHash = fundingOutput.extractHash();
+
+        if (fundingOutputHash.length == 20) {
+            // A 20-byte output hash is used by P2SH. That hash is constructed
+            // by applying OP_HASH160 on the redeem script. A 20-byte output
+            // hash is also the case for P2PKH and P2WPKH (OP_HASH160 on the
+            // public key). However, since we compare the actual output hash
+            // with an expected redeem script hash, the transaction type
+            // doesn't matter since the check will success only for P2SH
+            // with expected script hash value.
```

Comment:
20-byte hashes are not only for P2SH but also for P2PKH and P2WPKH and this is what I wrote in the description. And yes, in that case, we don't have to worry about anything else since we compare the actual hash and we make sure it's P2SH we expect and not an arbitrary P2(W)PKH. Just wanted to flag it explicitly in case we need to change something here. P2WSH, in turn, has a 32-byte hash as shown in the second condition.

Reply:
What about this:
```
// A 20-byte output hash is used by P2SH. That hash is constructed
// by applying OP_HASH160 on the unlock script. A 20-byte output
// hash is used as well by P2PKH and P2WPKH (OP_HASH160 on the
// public key). However, since we compare the actual output hash
// with an expected redeem script hash, this check will succeed only
// for P2SH transaction type with expected script hash value and for
// P2PKH and P2WPKH, it will fail on the output hash comparison with
// the expected redeem script hash.
```
?

Reply:
This is exactly what I'm saying :D

Reply:
Yes but this version is optimized for people aged over 30 👴🏻 👵🏻  It is even more explicit than yours 😆

Reply:
Sure, sure. Will change!

Reply:
See https://github.com/keep-network/tbtc-v2/pull/90/commits/bf19484ea615e888180e317570bf12613fcc34bb